### PR TITLE
verify: Added extra statistics for block index; Printable on debug; Interesting results

### PR DESCRIPTION
**Pull request by bwplotka**
Was playing around a couple of statistics of real block indexes and this is crazy!

For example one customer's block from our production at Red Hat has following stats:

```
ts=2020-10-30T16:05:15.969081066Z caller=level.go:63 level=debug verifiers=index_known_issues verifier=index_known_issues stats="{TotalSeries:8377876 OutOfOrderSeries:0 OutOfOrderChunks:0 DuplicatedChunks:0 OutsideChunks:0 CompleteOutsideChunks:0 Issue347OutsideChunks:0 OutOfOrderLabels:0 SeriesMinLifeDuration:0s SeriesAvgLifeDuration:13h7m1.399s SeriesMaxLifeDuration:311h50m40.379s SeriesMinLifeDurationWithoutSingleSampleSeries:14.91s SeriesAvgLifeDurationWithoutSingleSampleSeries:13h46m22.193s SeriesMaxLifeDurationWithoutSingleSampleSeries:311h50m40.379s SeriesMinChunks:1 SeriesAvgChunks:8 SeriesMaxChunks:165 TotalChunks:67874256 ChunkMinDuration:0s ChunkAvgDuration:1h37m8.646s ChunkMaxDuration:1h59m56.586s ChunkMinSize:23 ChunkAvgSize:135 ChunkMaxSize:1209 SingleSampleSeries:398902 SingleSampleChunks:1330558}" id=01DN3SK96XDAEKRB1AN30AAW6E
```

We can actually learn a lot here. This is a non-downsampled, 2-week block that has 8.3 Million of series.

Interesting things we can find from this :exploding_head:  :point_down: 

* Despite block being 2 weeks long, avg lifetime of series is just `SeriesAvgLifeDuration:13h7m1.399s`
* Biggest chunk is just over 1KB
* There are tons of 1 sample chunks and whole series with just 1 sample: `SingleSampleSeries:398902 SingleSampleChunks:1330558`.. 400k of 8.3M series have one sample, thus most likely are not usable / not graphable. This might mean some sneaky app using Prometheus as event-based system. :facepalm: 

Wonder what your blocs look like, so shipping this code so everyone can check. Even usable for blocks in Prometheus local disk e.g:

```
#!/usr/bin/env bash

thanos tools bucket verify --log.level debug -i "index_known_issues" --objstore.config "
type: FILESYSTEM
config:
  directory: /<prometheus_local_dir>
"
``` 

Is that healthy? @brian-brazil @pracucci @kakkoyun @s-urbaniak @gouthamve @codesome @smarterclayton ;p

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: cmd/thanos/config.go [hunk 1]
```diff
@@ -104,7 +104,7 @@ type reloaderConfig struct {
 
 func (rc *reloaderConfig) registerFlag(cmd extkingpin.FlagClause) *reloaderConfig {
 	cmd.Flag("reloader.config-file",
-		"Config file watched by the reloader.").
+		"Context file watched by the reloader.").
```
Comment (kakkoyun) [inline cmd/thanos/config.go:107]: I think this was unintended change.

File: pkg/objstore/azure/azure_test.go [hunk 2]
```diff
@@ -95,7 +95,7 @@ func TestConfig_validate(t *testing.T) {
 			}
 			err := conf.validate()
 			if (err != nil) != tt.wantErr {
-				t.Errorf("Config.validate() error = %v, wantErr %v", err, tt.wantErr)
+				t.Errorf("Context.validate() error = %v, wantErr %v", err, tt.wantErr)
```
Comment (kakkoyun) [inline pkg/objstore/azure/azure_test.go:98]: 🤔

File: pkg/verifier/verify.go [hunk 3]
```diff
@@ -21,16 +21,18 @@ import (
...
 type VerifierRepairer interface {
 	IssueID() string
-	VerifyRepair(ctx context.Context, conf Config, idMatcher func(ulid.ULID) bool, repair bool) error
+	VerifyRepair(ctx Context, idMatcher func(ulid.ULID) bool, repair bool) error
 }
 
-// Config is an verifier config.
-type Config struct {
+// Context is an verifier config.
+type Context struct {
```
Comment (kakkoyun) [inline pkg/verifier/verify.go:33]: 👍

**Review by kakkoyun — COMMENTED** [review]
Looks good. I think there are some unintended `Config`->`Context` transformations.

File: cmd/thanos/config.go (hunk 1 above)
Comment (bwplotka) [reply cmd/thanos/config.go:107]: uuuups

**Comment by bwplotka** [comment]
Fixed :hugs: Thanks @kakkoyun for super quick review

**Comment by s-urbaniak** [comment]
@bwplotka i think this is wonderful! :tada: questions: could this data be a) machine-retrievable? and b) available in the UI?

**Comment by bwplotka** [comment]
> could this data be a) machine-retrievable? and b) available in the UI?

a) you mean API? Why not, feel free to add an issue for it :hugs: (This API would need to download block index though)

b) Of what service? There is plan to have BucketViewer to have those features (cc @kunal-kushwaha) and to migrate such view to Prometheus

**Comment by pracucci** [comment]
This is very interesting. Thanks!

**Pull request merged**